certora/specs/ERC20FlashMint.spec

```diff
@@ -0,0 +1,43 @@
+import "helpers.spec"
+import "methods/IERC20.spec"
+import "methods/IERC3156.spec"
+
+methods {
+    // non standard ERC3156 functions
+    flashFeeReceiver() returns (address) envfree
+
+    // patched - function summaries below
+    _mint(address account, uint256 amount) => specMint(account, amount)
+    _burn(address account, uint256 amount) => specBurn(account, amount)
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Ghost: track mint and burns in the CVL                                                                              │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+ghost mapping(address => uint256) trackedMintAmount;
+ghost mapping(address => uint256) trackedBurnAmount;
+
+function specMint(address account, uint256 amount) returns bool { trackedMintAmount[account] = amount; return true; }
+function specBurn(address account, uint256 amount) returns bool { trackedBurnAmount[account] = amount; return true; }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: When doing a flashLoan, "amount" is minted and "amount + fee" is burnt                                        │
```

Comment:
It's not accurate to say that `amount + fee` is burnt, it's not true when the receiver is non-zero.

Should we also test that the fee is transferred to the receiver?

Reply:
Short answer is I'll try to address that. But there is A LOT I'd like to say about this rule so we should probably get on a call for that

Reply:
Lets try:

- This rule was written before we had a fee recipient, so the idea of transfer to the recipient was not included
- This rule was targeting a "bad" harness, the devs did not realize it but the fee they was always 0 for the FV point of view.
- The core of the flash loan can do anything. Including minting/burning/transfering ... so any check that is balance or supply related should fail. If it doesn't that means the FV is not working properly.
- The workaround Certora found is to "capture" the `burn` call, and register it locally. I duplicated that for the `mint`.
  - Technically, we could do that for transfer as well. That would capture the transfer call, and we would be able to check that such a transfer was indeed executed. 
  - However, that would require the `_transfer` function to be patched to public (I don't think we can do all that on internal methods). If we do so, all the ERC20 spec will have to consider this new function as an entrypoint. Which raise an even bigger question:  **what are we formally verifying ?**. Do we want to formally verify the behavior of the contract with regard to its public/external interface. Do we want to formally verify the internal interface as well?

Reply:
I was wrong about one thing: it is possible to summaries internal function. So no need to make `_transfer` public to check the transfer to the recipient.

Question of knowing if `_transfer` should be part of the FV spec remains open.